### PR TITLE
Refactor infrastructure extension management with DeployWaiter concept

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area testing
/kind enhancement
/priority normal

**What this PR does / why we need it**:
Similar to #2442, this PR is now refactoring the management of the `Infrastructure` extension resource as part of the `Botanist` structure in a way to leverage the new `DeployWaiter` interface introduced with #1980.

It also fixes a wrongly called method in the `WaitCleanup` function of the `Network` extension management.

/invite @mvladev 

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/infrastructure.go
Comment (swilen-iwanow) [inline pkg/operation/botanist/infrastructure.go:66]: I think this similarly to the Network will throw a nil pointer, because inside the restore we are trying to access some fields from the Infrastructure obj. May be we can temporally add  `GET` of the Infra object, until @plkokanov 
 change is merged.

Comment (rfranzke) [reply pkg/operation/botanist/infrastructure.go:66]: @swilen-iwanow hm, how does it work for `Network`s currently? There is no `GET` call either, or am I missing something?
https://github.com/gardener/gardener/blob/5773c96710adabe61574f39428ab2a6d43645a1f/pkg/operation/botanist/network.go#L55-L62

Comment (swilen-iwanow) [reply pkg/operation/botanist/infrastructure.go:66]: No. My point was that with the change of the `Network`s to use a component approach and passing an empty object to the `restoreExtensionObject`, instead of the created object like it used to be, the restore scenario for the `Network`s is not working anymore. With the Infra it will be the same. What I'm saying is that we can add a quick fix by getting the object inside the `restoreExtensionObject` func, which could be done in a separate PR.

Comment (rfranzke) [reply pkg/operation/botanist/infrastructure.go:66]: OK, understood, thanks. Can you open such a PR for the `Network`? I'm happy to incorporate the changes in this PR as well. Or, you wait until this PR is merged and do it together at once for both `Network` and `Infrastructure`.

Comment (swilen-iwanow) [reply pkg/operation/botanist/infrastructure.go:66]: I can wait until this PR is merged. If #2511 is merged as well I will just implement the new interfaces that are supposed to fix this

File: pkg/operation/botanist/extensions/infrastructure/infrastructure.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/infrastructure/infrastructure.go]: I guess it is unused.
```suggestion
```

File: pkg/operation/botanist/extensions/infrastructure/infrastructure_test.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/infrastructure/infrastructure_test.go]: You could simply use `.Return` in this case.

```suggestion
			c.EXPECT().
				Get(gomock.Any(), kutil.Key(namespace, name), &extensionsv1alpha1.Infrastructure{}).
				Return(apierrors.NewNotFound(schema.GroupResource{}, name)).AnyTimes()
```

File: pkg/operation/botanist/extensions/infrastructure/infrastructure_test.go
Comment (ialidzhikov) [inline pkg/operation/botanist/extensions/infrastructure/infrastructure_test.go]:
```suggestion
			c.EXPECT().
				Get(gomock.Any(), kutil.Key(namespace, name), infra).
				Return(apierrors.NewNotFound(schema.GroupResource{}, name))
```

File: pkg/gardenlet/controller/shoot/shoot_control_delete.go
Comment (ialidzhikov) [inline pkg/gardenlet/controller/shoot/shoot_control_delete.go]: Hm, but this will only delete the DNSEntry but won't wait until it is removed from store. Is it this change on purpose? I guess it is better to fail with `failed to wait until ... is deleted` instead of `failed to wait until Shoot namespace in the seed is deleted`.

Comment (rfranzke) [reply pkg/gardenlet/controller/shoot/shoot_control_delete.go]: Hm, agreed, I wasn't aware of these implicit consequences. I'll change it back... thanks for finding it.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**
